sandbox/modules/identity/src/main/java/org/opensearch/identity/SecurityFilter.java

```diff
@@ -0,0 +1,139 @@
+        this.client = client;
+        this.threadContext = threadPool.getThreadContext();
+        this.cs = cs;
+    }
+
+    @Override
+    public int order() {
+        return Integer.MIN_VALUE;
+    }
+
+    @Override
+    public <Request extends ActionRequest, Response extends ActionResponse> void apply(
+        Task task,
+        final String action,
+        Request request,
+        ActionListener<Response> listener,
+        ActionFilterChain<Request, Response> chain
+    ) {
+        try (StoredContext ctx = threadContext.newStoredContext(true)) {
+            org.apache.logging.log4j.ThreadContext.clearAll();
+            apply0(task, action, request, listener, chain);
+        }
+    }
+
+    private <Request extends ActionRequest, Response extends ActionResponse> void apply0(
+        Task task,
+        final String action,
+        Request request,
+        ActionListener<Response> listener,
+        ActionFilterChain<Request, Response> chain
+    ) {
+        try {
+            // TODO Get jwt here and verify
+            // TODO Move this logic to right after successful login
+            if (threadContext.getHeader(ThreadContextConstants.OPENSEARCH_AUTHENTICATION_TOKEN_HEADER) != null) {
+                String encodedJwt = threadContext.getHeader(ThreadContextConstants.OPENSEARCH_AUTHENTICATION_TOKEN_HEADER);
+                String prefix = "(nodeName="
+                    + cs.localNode().getId()
+                    + ", requestId="
+                    + request.getParentTask().getId()
+                    + ", action="
+                    + action
+                    + " apply0)";
+                log.info(prefix + " Access token provided " + encodedJwt);
+            } else {
+                // TODO Figure out where internal actions are invoked and create token on invocation
+                // No token provided, may be an internal request
+                // Token in ThreadContext is created on REST layer and passed to Transport Layer.
+                String prefix = "(nodeName="
+                    + cs.localNode().getName()
+                    + ", requestId="
+                    + request.getParentTask().getId()
+                    + ", action="
+                    + action
+                    + " apply0)";
+                log.info(prefix + "No authorization provided in the request, internal request");
+                // String err = "Access token not provided";
+                // listener.onFailure(new OpenSearchSecurityException(err, RestStatus.FORBIDDEN));
+            }
+
+            final PrivilegesEvaluatorResponse pres = new PrivilegesEvaluatorResponse(); // eval.evaluate(user, action, request, task,
+                                                                                        // injectedRoles);
+            pres.allowed = true;
+
+            if (log.isDebugEnabled()) {
+                log.debug(pres.toString());
+            }
+
+            if (pres.isAllowed()) {
```

Comment:
this check would always be true based on setting on line 99. Is that expected or should line 99 be changed to something conditional?

Reply:
That is more boilerplate that is meant to be expanded on when authorization is tackled. When I started working on this branch one of the big ideas was to deprecate the usage of serializing the User object and transporting around the cluster on the thread context headers as a proxy for a request that has previously been authenticated on the rest layer. Instead of serializing the user object, this PR introduces tokens for authentication information and transports them around the cluster using the thread context headers. I originally wanted to terminate an internal transport request if it did not contain authentication information with it. The problem I faced is that this header would be populated for internal transport requests that originated from the Rest Layer, but it was not present for internal actions invoked from elsewhere.

In the future PR we need to investigate where internal actions are invoked from and choose whether we should send information indicating that its starting from an internal source.